rewindable-index/src/RewindableIndex/Storable.hs

```diff
@@ -288,28 +288,28 @@ rewind
      Rewindable h
   => HasPoint (StorableEvent h) (StorablePoint h)
   => PrimMonad (StorableMonad h)
-  => Eq (StorablePoint h)
+  => Ord (StorablePoint h)
   => StorablePoint h
   -> State h
   -> StorableMonad h (Maybe (State h))
-rewind p s = do
-  m' <- rewindMemory
-  h' <- rewindStorage p (s ^. handle)
-  -- The implementation here is a little non-trivial. If the rollback point is in memory
-  -- then we don't need to rewind the disk. If we need to rewind to some point stored
-  -- on disk, then the memory needs to be reset.
-  pure $ m' <|> resetMemory s <$> h'
-  where
-    rewindMemory :: StorableMonad h (Maybe (State h))
-    rewindMemory = do
-      v <- V.freeze $ VM.slice 0 (s ^. storage . cursor) (s ^. storage . events)
-      pure $ do
-        ix   <- VG.findIndex (\e -> getPoint e == p) v
-        pure $ s & storage . cursor .~ (ix + 1)
-    resetMemory :: State h -> h -> State h
-    resetMemory s' h =
-      s' & handle  .~ h
-         & storage . cursor .~ 0
+rewind p s = if s ^. storage . cursor == 0
+  -- Buffer is empty, rewind storage just in case:
```

Comment:
What is the difference between this code and the one before it (other than the change from `=` to `<` and `ix + 1` to `ix`)? Wouldn't the previous code work with just those changes?

Reply:
The difference is that we don't rewind storage when it is not required.

I.e, with just the changes you mention, `findIndex` would return `Nothing` in both when buffer is empty or when all events are older than the rewind point. In the first case storage needs to be rewound, in the second case since we know that all events in buffer are older than rewind point then everything in storage must be as well.

Also, with just the changes you mention, `findIndex` would return `Just ix` in both when an event with point later than rewind point is the *first element* in the buffer or when it's somewhere in the middle. If the found index is the first point in buffer then storage needs to be rewound, in the second case only the buffer needs to be rewound as there are events in the buffer that are older than the rewind point -- which means events in storage are even older still.